src/coreclr/jit/promotionliveness.cpp

```diff
@@ -0,0 +1,869 @@
+//   the rest of the JIT for a short while between the liveness is computed and
+//   the replacement phase has run: in particular, after this liveness pass you
+//   may see a node like:
+//
+//       LCL_FLD   int    V16 tmp9         [+8] (last use)
+//
+//   that indicates that this particular field (or the remainder if it wasn't
+//   promoted) is dying, not that V16 itself is dying. After replacement has
+//   run the semantics align with the rest of the JIT: in the promoted case V16
+//   [+8] would be replaced by its promoted field local, and in the remainder
+//   case all non-remainder uses of V16 would also be.
+//
+//   There is one catch which is struct uses of the local. These can indicate
+//   deaths of multiple fields and also the remainder, so this information is
+//   stored on the side. PromotionLiveness::GetDeathsForStructLocal is used to
+//   query this information.
+//
+//   The liveness information is used by decomposition to avoid creating dead
+//   stores, and also to mark the replacement field uses/defs with proper
+//   up-to-date liveness information to be used by future phases (forward sub
+//   and morph, as of writing this). It is also used to avoid creating
+//   unnecessary read-backs; this is mostly just a TP optimization as future
+//   liveness passes would be expected to DCE these anyway.
+//
+//   Avoiding the creation of dead stores to the remainder is especially
+//   important as these otherwise would often end up looking like partial
+//   definitions, and the other liveness passes handle partial definitions very
+//   conservatively and are not able to DCE them.
+//
+//   Unlike the other liveness passes we keep the per-block liveness
+//   information on the side and we do not update BasicBlock::bbLiveIn et al.
+//   This relies on downstream phases not requiring/wanting to use per-basic
+//   block live-in/live-out/var-use/var-def sets. To be able to update these we
+//   would need to give the new locals "regular" tracked indices (i.e. allocate
+//   a lvVarIndex).
+//
+//   The indices allocated and used internally within the liveness computation
+//   are "dense" in the sense that the bit vectors only have indices for
+//   remainders and the replacement fields introduced by this pass. In other
+//   words, we allocate 1 + num_fields indices for each promoted struct local).
+//
+void PromotionLiveness::Run()
+{
+    m_structLclToTrackedIndex = new (m_compiler, CMK_Promotion) unsigned[m_aggregates.size()]{};
+    unsigned trackedIndex     = 0;
+    for (size_t lclNum = 0; lclNum < m_aggregates.size(); lclNum++)
+    {
+        AggregateInfo* agg = m_aggregates[lclNum];
+        if (agg == nullptr)
+        {
+            continue;
+        }
+
+        m_structLclToTrackedIndex[lclNum] = trackedIndex;
+        // TODO: We need a scalability limit on these, we cannot always track
+        // the remainder and all fields.
```

Comment:
Will address this in a separate follow-up.

Reply:
This is similar to the "coarsening" discussion we had re alias analysis the other day, you should be able to have models where you group together fields and or fields + remainder and get approximate liveness for them.

Would be interesting to know if you see cases where there are large structs with "hot fields" so that it is worth tracking a subset but coarsening everything else, and if that diverges somehow from just promoting the hot fields and leaving the rest.

Reply:
> Would be interesting to know if you see cases where there are large structs with "hot fields" so that it is worth tracking a subset but coarsening everything else, and if that diverges somehow from just promoting the hot fields and leaving the rest.

That's a good point... the heuristics should likely include something to avoid creating locals that are just going to end up untracked.